grr/server/grr_response_server/gui/ui/components/expandable_hash/expandable_hash_test.ts

```diff
@@ -27,32 +36,75 @@ describe('ExpandableHash component', () => {
     fixture.detectChanges();
   });
 
-  function getDisplayedHashText() {
-    const hashDiv = fixture.debugElement.query(By.css('.displayed-text'));
-    return hashDiv.nativeElement.innerText;
+  class ExpandableHashDOM {
+    expandButton = this.root.query(By.css('.button-expand'));
+    expandButtonText = this.expandButton?.nativeElement.innerText;
+
+    noHashesSpan = this.root.query(By.css('.no-hashes'));
+    noHashesSpanText = this.noHashesSpan?.nativeElement.innerText;
+
+    constructor(readonly root: DebugElement) { }
   }
 
-  function initComponentWithHashes(hashes: Hash) {
+  function initComponentWithHashes(hashes: Hash): ExpandableHashDOM {
     component.hashes = hashes;
     fixture.detectChanges();
+    return new ExpandableHashDOM(fixture.debugElement);
   }
 
-  it('should display en dash (–) if no hashes are available', () => {
+  it('should only display an en dash (–) if no hashes are available', () => {
     const noHashes = { };
     const emDash = '–';
 
-    initComponentWithHashes(noHashes);
+    const DOM = initComponentWithHashes(noHashes);
+
+    expect(DOM.expandButton).toBeFalsy();
 
-    expect(getDisplayedHashText()).toBe(emDash);
+    expect(DOM.noHashesSpan).toBeTruthy();
+    expect(DOM.noHashesSpanText).toBe(emDash);
   });
 
-  it('should display text if 1 hash is available', () => {
+  it('should only display \"Copy value\" text if 1 hash is available', () => {
     const oneHash = {
       sha256: 'sha256',
     };
 
     initComponentWithHashes(oneHash);
 
-    expect(getDisplayedHashText()).toBe('Copy value');
+    const DOM = initComponentWithHashes(oneHash);
+
+    expect(DOM.noHashesSpan).toBeFalsy();
+
+    expect(DOM.expandButton).toBeTruthy();
+    expect(DOM.expandButtonText).toBe('Copy value');
+  });
+});
+
+describe('HashTextAggregator', () => {
```

Comment:
You test the function of HashTextAggregator (which could be inlined, see above), but not that the actual website displays the information you want, if I did not miss something.

Please test that:
1) The menu shows all hashes.
2) The menu shows the (prefix) of the correct hash digest.
3) Clicking actually copies the expected value (see approval_test.ts for an example on testing the clipboard cdk).

Reply:
I will do that in the Selenium tests, as I need to click on the menu in order to expand it and show those.

Reply:
Selenium tests are generally slow and hard to introspect. They are good for big integration tests, but don't perfectly fit testing single components. Check out https://material.angular.io/guide/using-component-harnesses and Angular's `MatMenuHarness` to test this in a unit test.